File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/ComposingWith.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.Fragment;
import dev.typr.foundations.OperationRead;
import dev.typr.foundations.PgTypes;
import dev.typr.foundations.RowCodec;
import dev.typr.foundations.Transactor;
import java.util.List;

@SuppressWarnings("unused")
public class ComposingWith {
  // start
  record User(int id, String name) {}

  record Order(int id, int userId, String product) {}

  record Dashboard(long userCount, List<Order> recentOrders) {}

  record Stats(long userCount, long orderCount, long revenue) {}

  static RowCodec<Order> orderCodec =
      RowCodec.<Order>builder()
          .field(PgTypes.int4, Order::id)
          .field(PgTypes.int4, Order::userId)
          .field(PgTypes.text, Order::product)
          .build(Order::new);

  Transactor tx = null; // placeholder

  // Combine two independent queries — both run in one transaction
  OperationRead<Long> countUsers =
      Fragment.of("SELECT count(*) FROM users").query(RowCodec.of(PgTypes.int8).exactlyOne());
  OperationRead<List<Order>> recentOrders =
      Fragment.of(
              """
              SELECT * FROM orders
              ORDER BY id DESC LIMIT 10\
              """)
          .query(orderCodec.all());

  Dashboard dashboard() {
    return countUsers.combineWith(recentOrders, Dashboard::new).transactRead(tx);
  }

  // Three-way: all run in one transaction, results combined
  OperationRead<Long> countOrders =
      Fragment.of("SELECT count(*) FROM orders").query(RowCodec.of(PgTypes.int8).exactlyOne());
  OperationRead<Long> totalRevenue =
      Fragment.of(
              """
              SELECT coalesce(sum(amount), 0)
              FROM orders\
              """)
          .query(RowCodec.of(PgTypes.int8).exactlyOne());

  Stats stats() {
    return countUsers.combineWith(countOrders, totalRevenue, Stats::new).transactRead(tx);
  }
  // stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/core/ComposingWith.scala
package dev.typr.foundationssc.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.Fragment.sql
import dev.typr.foundationssc.data.*

@SuppressWarnings(Array("unused"))
object ComposingWith:
  // start
  case class User(id: Int, name: String)
  case class Order(id: Int, userId: Int, product: String)
  case class Dashboard(userCount: Long, recentOrders: List[Order])
  case class Stats(userCount: Long, orderCount: Long, revenue: Long)

  val orderCodec: RowCodec[Order] = RowCodec
    .builder[Order]()
    .field(PgTypes.int4)(_.id)
    .field(PgTypes.int4)(_.userId)
    .field(PgTypes.text)(_.product)
    .build(Order.apply)

  var tx: Transactor = null // placeholder

  // Combine two independent queries in one transaction
  val countUsers: OperationRead[Long] =
    sql"SELECT count(*) FROM users"
      .query(RowCodec.of(PgTypes.int8).exactlyOne())
  val recentOrders: OperationRead[List[Order]] =
    sql"SELECT * FROM orders ORDER BY id DESC LIMIT 10"
      .query(orderCodec.all())

  def dashboard(): Dashboard =
    countUsers
      .combineWith(recentOrders)(Dashboard.apply)
      .transact(tx)

  // Three-way: all run in one transaction
  val countOrders: OperationRead[Long] =
    sql"SELECT count(*) FROM orders"
      .query(RowCodec.of(PgTypes.int8).exactlyOne())
  val totalRevenue: OperationRead[Long] =
    sql"SELECT coalesce(sum(amount), 0) FROM orders"
      .query(RowCodec.of(PgTypes.int8).exactlyOne())

  def stats(): Stats =
    countUsers
      .combineWith(countOrders, totalRevenue)(Stats.apply)
      .transact(tx)
  // stop
